### PR TITLE
Initialize api tokens index if it doesn't exist 

### PR DESCRIPTION
**Pull request by derek-ho**
### Description
Initializes a new index for API tokens if it doesn't exist

### Issues Resolved
Closes: https://github.com/opensearch-project/security/issues/4900

Is this a backport? If so, please add backport PR # and/or commits #, and remove `backport-failed` label from the original PR.

Do these changes introduce new permission(s) to be displayed in the static dropdown on the front-end? If so, please open a draft PR in the security dashboards plugin and link the draft PR here

### Testing
[Please provide details of testing done: unit testing, integration testing and manual testing]

### Check List
- [x] New functionality includes testing
- [ ] New functionality has been documented
- [ ] New Roles/Permissions have a corresponding security dashboards plugin PR
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md)
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/security/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java [hunk 1]
```diff
@@ -431,6 +431,32 @@ Future<Void> executeConfigurationInitialization(final SecurityMetadata securityM
         return CompletableFuture.completedFuture(null);
     }
 
+    public void createApiTokenIndex() {
+        try {
+            // wait for the cluster here until it will finish managed node election
+            while (clusterService.state().blocks().hasGlobalBlockWithStatus(RestStatus.SERVICE_UNAVAILABLE)) {
```
[the remark this answers is not in the document]
Comment (derek-ho) [reply src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:437]: This is necessary as without it we could try to create an index before the cluster is ready to handle this request. @DarshitChanpura @cwperks Do you folks think we still need this check though? As that check should already be performed by the creation of the `opendistro_security` index, depending on the value of `"plugins.security.allow_default_init_securityindex"` setting. In any case I was just taking inspiration from the creation of the `opendistro_security` index.

File: src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java [hunk 2]
```diff
@@ -431,6 +431,32 @@ Future<Void> executeConfigurationInitialization(final SecurityMetadata securityM
...
+                LOGGER.info("Wait for cluster to be available ...");
+                TimeUnit.SECONDS.sleep(1);
+            }
+
+            try {
+                final ThreadContext threadContext = threadPool.getThreadContext();
+                try (StoredContext ctx = threadContext.stashContext()) {
```
[the remark this answers is not in the document]
Comment (derek-ho) [reply src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:444]: Like the above, this is taken from the creation of `opendistro_security` index, which I assume has this to handle edge cases since the security index may not be initialized yet. Again not completely sure if its needed but I think it is fine?

File: src/test/java/org/opensearch/security/configuration/ConfigurationRepositoryTest.java [hunk 3]
```diff
@@ -148,6 +152,23 @@ public void initOnNodeStart_withSecurityIndexNotCreatedShouldNotSetInstallDefaul
         assertThat(result.join(), is(false));
     }
 
+    @Test
+    public void createApiTokenIndex_createsApiTokenIndex() {
```
[the remark this answers is not in the document]
Comment (derek-ho) [reply src/test/java/org/opensearch/security/configuration/ConfigurationRepositoryTest.java]: Renamed in: [fc13790](https://github.com/opensearch-project/security/pull/4899/commits/fc1379030cbe6016878cc3aac66156a7ff2f1dd8)

File: src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java [hunk 4]
```diff
@@ -201,8 +202,8 @@ private void initalizeClusterConfiguration(final boolean installDefaultConfig) {
                         try (StoredContext ctx = threadContext.stashContext()) {
                             threadContext.putHeader(ConfigConstants.OPENDISTRO_SECURITY_CONF_REQUEST_HEADER, "true");
 
-                            createSecurityIndexIfAbsent();
-                            waitForSecurityIndexToBeAtLeastYellow();
+                            createIndexIfAbsent(securityIndex);
```
Comment (cwperks) [inline src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:205]: I think this class should be reserved for initializing the security index. The security index will always need to be created on bootstrap.

For most plugins, I believe they create a system index when its needed if it doesn't exist instead of creating it on bootstrap. i.e. A user tries to issue an auth token and security tries to store the metadata, but the index doesn't exist so create it at that time. 

See this class from AD plugin: https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/ad/indices/ADIndexManagement.java

Comment (derek-ho) [reply src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:205]: Sure this approach makes sense too. I also thought of this approach at first, but thought that it will add too much overhead to try and create this index for every request if it doesn't exist. However, we can limit this to index operations (create, revoke, etc.), and do not need to perform this for every authC request. I will try to do it this way instead.

**Comment by derek-ho** [comment]
Closing in favor of: https://github.com/opensearch-project/security/pull/4912/files
